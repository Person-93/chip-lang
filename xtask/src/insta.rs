use crate::{CommandExt, ErrorExt, Parser, ResultExt};
use std::{env, process::Command};

/// Run `cargo nextest run` then `cargo insta review`
#[derive(Default, Parser)]
pub struct Cli {}

pub fn main(Cli {}: Cli) -> ! {
  cargo()
    .args(["nextest", "run"])
    .env("INSTA_FORCE_PASS", "1")
    .wait()
    .exit_on_err();
  cargo().args(["insta", "review"]).exec().exit();
}

fn cargo() -> Command {
  env::var_os("CARGO").map_or_else(|| Command::new("cargo"), Command::new)
}
