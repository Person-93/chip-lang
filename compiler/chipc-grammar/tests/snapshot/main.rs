use crate::debug_print::print;
use insta::{assert_snapshot, glob};
use std::fs;
use tree_sitter::Parser;

fn main() {
  let mut parser = Parser::new();
  parser
    .set_language(chipc_grammar::language())
    .expect("assign language to parser");

  glob!("cases/*.chip", |path| {
    parser.reset();
    let source = fs::read_to_string(path).expect("read snapshot test case");
    let tree = parser
      .parse(&source, None)
      .expect("parse snapshot test case");
    assert_snapshot!(print(&tree, &source));
  });
}

mod debug_print {
  use std::fmt::{self, Write};
  use tree_sitter::{Point, Tree, TreeCursor};

  pub fn print(tree: &Tree, source: &str) -> String {
    let mut output = String::new();
    let mut cursor = tree.walk();
    output
      .print_node(&mut cursor, source, 0)
      .expect("debug format tree");
    output
  }

  impl<T: Write> NodePrinter for T {}

  trait NodePrinter: Write {
    fn print_node(
      &mut self,
      cursor: &mut TreeCursor,
      source: &str,
      indentation: u8,
    ) -> fmt::Result {
      let node = cursor.node();
      let kind = node.kind();
      let Point { row, column } = node.start_position();
      let Point {
        row: end_row,
        column: end_column,
      } = node.end_position();

      self.indent(indentation)?;
      write!(self, "{kind} @ {row}:{column}..{end_row}:{end_column}")?;
      if node.is_named() && node.child_count() == 0 {
        write!(self, "\t{}", node.utf8_text(source.as_bytes()).unwrap())?;
      }
      writeln!(self)?;

      if cursor.goto_first_child() {
        self.print_node(&mut cursor.clone(), source, indentation + 1)?;
        while cursor.goto_next_sibling() {
          self.print_node(&mut cursor.clone(), source, indentation + 1)?;
        }
      }

      Ok(())
    }

    fn indent(&mut self, indentation: u8) -> fmt::Result {
      for _ in 0..indentation {
        write!(self, "  ")?;
      }
      Ok(())
    }
  }
}
