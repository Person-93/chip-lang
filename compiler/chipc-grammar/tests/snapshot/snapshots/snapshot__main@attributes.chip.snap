---
source: compiler/chipc-grammar/tests/snapshot/main.rs
assertion_line: 18
expression: "print(&tree, &source)"
input_file: compiler/chipc-grammar/tests/snapshot/cases/attributes.chip
---
file @ 0:0..10:0
  #^[ @ 0:0..0:3
  attr @ 0:3..0:9
    path @ 0:3..0:9
      path_segment @ 0:3..0:9
        identifier @ 0:3..0:9	simple
  ] @ 0:9..0:10
  #^[ @ 1:0..1:3
  attr @ 1:3..1:22
    path @ 1:3..1:22
      path_segment @ 1:3..1:7
        identifier @ 1:3..1:7	attr
      :: @ 1:7..1:9
      path_segment @ 1:9..1:13
        identifier @ 1:9..1:13	with
      :: @ 1:13..1:15
      path_segment @ 1:15..1:16
        identifier @ 1:15..1:16	a
      :: @ 1:16..1:18
      path_segment @ 1:18..1:22
        identifier @ 1:18..1:22	path
  ] @ 1:22..1:23
  #^[ @ 2:0..2:3
  attr @ 2:3..2:19
    path @ 2:3..2:9
      path_segment @ 2:3..2:9
        identifier @ 2:3..2:9	string
    = @ 2:10..2:11
    string_literal @ 2:12..2:19	"hello"
  ] @ 2:19..2:20
  #^[ @ 3:0..3:3
  attr @ 3:3..3:9
    path @ 3:3..3:4
      path_segment @ 3:3..3:4
        identifier @ 3:3..3:4	n
    = @ 3:5..3:6
    numeric_literal @ 3:7..3:9
      digits @ 3:7..3:9	42
  ] @ 3:9..3:10
  #^[ @ 4:0..4:3
  attr @ 4:3..4:11
    path @ 4:3..4:4
      path_segment @ 4:3..4:4
        identifier @ 4:3..4:4	b
    = @ 4:5..4:6
    attr @ 4:7..4:11
      path @ 4:7..4:11
        path_segment @ 4:7..4:11
          identifier @ 4:7..4:11	true
  ] @ 4:11..4:12
  #^[ @ 5:0..5:3
  attr @ 5:3..5:31
    path @ 5:3..5:7
      path_segment @ 5:3..5:7
        identifier @ 5:3..5:7	call
    ( @ 5:7..5:8
    string_literal @ 5:8..5:14	"arg0"
    , @ 5:14..5:15
    attr @ 5:16..5:20
      path @ 5:16..5:20
        path_segment @ 5:16..5:20
          identifier @ 5:16..5:20	arg1
    , @ 5:20..5:21
    numeric_literal @ 5:22..5:24
      digits @ 5:22..5:24	42
    , @ 5:24..5:25
    attr @ 5:26..5:31
      path @ 5:26..5:31
        path_segment @ 5:26..5:31
          identifier @ 5:26..5:31	false
    ) @ 5:31..5:31
  ] @ 5:31..5:32
  #^[ @ 6:0..6:3
  attr @ 6:3..6:40
    path @ 6:3..6:7
      path_segment @ 6:3..6:7
        identifier @ 6:3..6:7	call
    ( @ 6:7..6:8
    attr @ 6:8..6:11
      path @ 6:8..6:11
        path_segment @ 6:8..6:11
          identifier @ 6:8..6:11	arg
    , @ 6:11..6:12
    attr @ 6:13..6:25
      path @ 6:13..6:19
        path_segment @ 6:13..6:19
          identifier @ 6:13..6:19	nested
      ( @ 6:19..6:20
      attr @ 6:20..6:24
        path @ 6:20..6:24
          path_segment @ 6:20..6:24
            identifier @ 6:20..6:24	call
      ) @ 6:24..6:25
    , @ 6:25..6:26
    attr @ 6:27..6:40
      path @ 6:27..6:31
        path_segment @ 6:27..6:31
          identifier @ 6:27..6:31	this
      = @ 6:32..6:33
      string_literal @ 6:34..6:40	"that"
    ) @ 6:40..6:40
  ] @ 6:40..6:41
  function @ 8:0..9:22
    #[ @ 8:0..8:2
    attr @ 8:2..8:11
      path @ 8:2..8:11
        path_segment @ 8:2..8:11
          identifier @ 8:2..8:11	some_attr
    ] @ 8:11..8:12
    fn @ 9:0..9:2
    identifier @ 9:3..9:16	fn_with_attrs
    ( @ 9:16..9:17
    ) @ 9:17..9:18
    : @ 9:18..9:19
    codeblock @ 9:20..9:22
      { @ 9:20..9:21
      } @ 9:21..9:22
